### PR TITLE
Feature 1788 ssidx

### LINKED ISSUE #1031: Enhance MET to compute the CBS score.
## Describe the New Feature ##
The AF project needs to generate the CBS score on its GPP verification. This requires adding another method of interpolation (areal). Michelle Harrold has provided some code (from Bob Craig) and helpful references (see attachments).

From Bob:

"All, the attached document describes the CBS verification process , I.7-36 Table F. The verification is done over a 1.5 degree remap of the model using an area averaging technique. The ground truth is the model analysis. I have attached the program we use to average the data. The index based on the CBS score is from UKMO. I attached the program that calculates it along with the configuration files."

I have attached the WMO document with I.7-36 Table F as well as the the areal weighting IDL script (around line 251 is the weighting code) and the CBS score python program. Also, the areal weighting script references a paper, which might have some helpful background:

https://www.wmo.int/pages/prog/www/BAS/documents/AnnexDoc3-3.pdf

The work for this issue may include supporting a new interpolation method, computing the CBS score, and writing its output somewhere. Recommend coordinating with @michelleharrold since she's likely familiar with the details.

### Acceptance Testing ###
*List input data types and sources.*
*Describe tests required for new functionality.*

### Time Estimate ###
*Estimate the amount of work required here.*
*Issues should represent approximately 1 to 3 days of work.*

### Sub-Issues ###
Consider breaking the new feature down into sub-issues.
- [ ] *Add a checkbox for each sub-issue here.*

### Relevant Deadlines ###
*List relevant project deadlines here or state NONE.*

### Funding Source ###
*Define the source of funding and account keys here or state NONE.*

### Assignee ###
- [x] Select **engineer(s)** or **no engineer** required: John HG
- [ ] Select **scientist(s)** or **no scientist** required: unknown

### Labels ###
- [x] Select **component(s)**
- [x] Select **priority**
- [x] Select **requestor(s)**

### Projects and Milestone ###
- [x] Review **projects** and select relevant **Repository** and **Organization** ones or add "alert:NEED PROJECT ASSIGNMENT" label
- [x] Select **milestone** to next major version milestone or "Future Versions"

## Define Related Issue(s) ##
Consider the impact to the other METplus components.
- [ ] [METplus](https://github.com/dtcenter/METplus/issues/new/choose), [MET](https://github.com/dtcenter/MET/issues/new/choose), [METdatadb](https://github.com/dtcenter/METdatadb/issues/new/choose), [METviewer](https://github.com/dtcenter/METviewer/issues/new/choose), [METexpress](https://github.com/dtcenter/METexpress/issues/new/choose), [METcalcpy](https://github.com/dtcenter/METcalcpy/issues/new/choose), [METplotpy](https://github.com/dtcenter/METplotpy/issues/new/choose)
Depending on the changes to the output, may need an issue for METviewer to handle new output. May also need a METplus issue if new configuration file options are added.

## New Feature Checklist ##
See the [METplus Workflow](https://dtcenter.github.io/METplus/Contributors_Guide/github_workflow.html) for details.
- [x] Complete the issue definition above, including the **Time Estimate** and **Funding source**.
- [x] Fork this repository or create a branch of **develop**.
Branch name: `feature_<Issue Number>_<Description>`
- [ ] Complete the development and test your changes.
- [x] Add/update log messages for easier debugging.
- [x] Add/update unit tests.
- [x] Add/update documentation.
- [x] Push local changes to GitHub.
- [x] Submit a pull request to merge into **develop**.
Pull request: `feature <Issue Number> <Description>`
- [x] Define the pull request metadata, as permissions allow.
Select: **Reviewer(s)**, **Project(s)**, **Milestone**, and **Linked issues**
- [x] Iterate until the reviewer(s) accept and merge your changes.
- [x] Delete your fork or branch.
- [x] Close this issue.

### LINKED ISSUE #1788: Enhance Stat-Analysis to write the GO Index and CBS Index into a new SSIDX STAT line type.
## Describe the Enhancement ##
This enhancement is recommended for the DTCenter Air Force LIS project. The Stat-Analysis tool supports the skill score index and GO index job types. The GO Index is just a special version of the skill score index job. As of MET version 10.0.0, Stat-Analysis simply prints the output for this job to the screen or to an output file. This task is to create a new .stat line type to store the relevant output.

Still need to decide exactly what columns should be written to this new line type. Could be as little as the final score, or as complex, as the individual scores/weights for each component. So this may end up being a fixed number of columns or variable length and based on the number of user-defined components in the configurable Skill Score index.

Be sure to define related issues for METdatadb and METviewer, as noted below.

### Time Estimate ###
3 days.

### Sub-Issues ###
Consider breaking the enhancement down into sub-issues.
No sub-issues, but this work will also satisfy a related GitHub issue: dtcenter/MET#1031

### Relevant Deadlines ###
October 2021

### Funding Source ###
2712221

### Assignee ###
- [x] Select **engineer(s)** or **no engineer** required: John HG
- [x] Select **scientist(s)** or **no scientist** required: Michelle and/or Tara

### Labels ###
- [x] Select **component(s)**
- [x] Select **priority**
- [x] Select **requestor(s)**

### Projects and Milestone ###
- [x] Review **projects** and select relevant **Repository** and **Organization** ones or add "alert:NEED PROJECT ASSIGNMENT" label
- [x] Select **milestone** to next major version milestone or "Future Versions"

## Define Related Issue(s) ##
Consider the impact to the other METplus components.
- [x] [METplus](https://github.com/dtcenter/METplus/issues/new/choose), [MET](https://github.com/dtcenter/MET/issues/new/choose), [METdatadb](https://github.com/dtcenter/METdatadb/issues/new/choose), [METviewer](https://github.com/dtcenter/METviewer/issues/new/choose), [METexpress](https://github.com/dtcenter/METexpress/issues/new/choose), [METcalcpy](https://github.com/dtcenter/METcalcpy/issues/new/choose), [METplotpy](https://github.com/dtcenter/METplotpy/issues/new/choose)
- [x] No METplus issue is needed here in my opinion.
- [x] Create a METdatadb issue to load this new line type: dtcenter/METdatadb#63
- [x] Create a METviewer issue to plot these statistics via the batch and GUI: dtcenter/METviewer#327

## Enhancement Checklist ##
See the [METplus Workflow](https://dtcenter.github.io/METplus/Contributors_Guide/github_workflow.html) for details.
- [x] Complete the issue definition above, including the **Time Estimate** and **Funding Source**.
- [x] Fork this repository or create a branch of **develop**.
Branch name: `feature_<Issue Number>_<Description>`
- [x] Complete the development and test your changes.
- [x] Add/update log messages for easier debugging.
- [x] Add/update unit tests.
- [x] Add/update documentation.
- [x] Push local changes to GitHub.
- [x] Submit a pull request to merge into **develop**.
Pull request: `feature <Issue Number> <Description>`
- [x] Define the pull request metadata, as permissions allow.
Select: **Reviewer(s)**, **Project(s)**, **Milestone**, and **Linked issues**
- [x] Iterate until the reviewer(s) accept and merge your changes.
- [x] Delete your fork or branch.
- [x] Close this issue.

### PR DESCRIPTION
**Pull request by JohnHalleyGotway**
## Pull Request Testing ##

- [x] Describe testing already performed for these changes:</br>
Manually tested these changes extensively, running go_index, cbs_index, and custom ss_index jobs both on the command line and via STATAnalysisConfig files. Updated unit_ref_config.xml with new calls to STAT-Analysis to exercise these changes.

- [x] Recommend testing for the reviewer(s) to perform, including the location of input datasets, and any additional instructions:</br>
Please see this comment for direction on where/what to test:
https://github.com/dtcenter/MET/issues/1788#issuecomment-906595599

- [x] Do these changes include sufficient documentation updates, ensuring that no errors or warnings exist in the build of the documentation? **[Yes]**
Major changes and clarifications to the STAT-Analysis chapter of the User's Guide.

- [x] Do these changes include sufficient testing updates? **[Yes]**
New calls to STAT-Analysis in unit_ref_config.xml to test the -go_index and a custom -ss_index job. None actually call -cbs_index because I didn't want to increase the amount of input data, but I did manually test that specific one.

- [x] Will this PR result in changes to the test suite? **[Yes]**</br>
If **yes**, describe the new output and/or changes to the existing output:</br>
New STAT-Analysis output from unit_ref_config.xml.

- [x] Please complete this pull request review by **[8/30/21]**.</br>
I'd like to include this in MET version 10.1.0 beta2, scheduled for 8/31/21.

## Pull Request Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [x] Complete the PR definition above.
- [x] Ensure the PR title matches the feature or bugfix branch name.
- [x] Define the PR metadata, as permissions allow.
Select: **Reviewer(s)**
Select: **Organization** level software support **Project** or **Repository** level development cycle **Project**
Select: **Milestone** as the version that will include these changes
- [x] After submitting the PR, select **Linked issues** with the original issue number.
- [x] After the PR is approved, merge your changes. If permissions do not allow this, request that the reviewer do the merge.
- [x] Close the linked issue and delete your feature or bugfix branch from GitHub.

### CODE REVIEW

File: met/docs/Users_Guide/stat-analysis.rst [hunk 1]
```diff
@@ -52,10 +52,49 @@ When aggregating the matched pair line type (MPR) and computing an output contin
...
+
+In general, a skill score index is computed over several terms and the number and definition of those terms is configurable. It is computed by aggregating the output from earlier runs of the Point-Stat and/or Grid-Stat tools over one or more cases. When configuring a skill score index job, the following requirements apply:
+
+1. Exactly two models names must be chosen. The first is interpreted as the forecast model and the second is the reference model, against which the performance of the forecast should be measured. Specify this with the "model" configuration file entry or using the "-model" job command option.
+
+2. The forecast variable name, level, lead time, line type, column, and weight options must be specified. If the value remains contant for all the terms, set its to an array of length one. If the value changes for at least one term, specify an array entry for each term. Specify these with the "fcst_var", "fcst_lev", "lead_time", "line_type", "column", and "weight" configuration file entries, respectively, or use the corresponding job command options.
```
Comment (jprestop) [inline met/docs/Users_Guide/stat-analysis.rst]:
```suggestion
2. The forecast variable name, level, lead time, line type, column, and weight options must be specified. If the value remains constant for all the terms, set it to an array of length one. If the value changes for at least one term, specify an array entry for each term. Specify these with the "fcst_var", "fcst_lev", "lead_time", "line_type", "column", and "weight" configuration file entries, respectively, or use the corresponding job command options.
```

File: met/docs/Users_Guide/stat-analysis.rst [hunk 2]
```diff
@@ -52,10 +52,49 @@ When aggregating the matched pair line type (MPR) and computing an output contin
...
+
+3. While these line types are required, additional options may be provided for each term, including the observation type ("obtype"), verification region ("vx_mask"), and interpolation method ("interp_mthd"). Specify each as single value or provide a value for each term.
+
+4. Only the SL1L2 and CTC input line types are supported, and the input Point-Stat and/or Grid-Stat output must contain these line types.
+
+5. For the SL1L2 line type, set the "column" entry to the CNT output column that contains the statistic of interest (e.g. RMSE for root-mean-squared-error). Note, only those continuous statistics that are derivable from SL1L2 lines can be used.
+
+6. For the CTC line type, set the "column" entry to the CTS output column that contains the statistic of intereest (.e.g. PODY for probability of detecting yes). Note, consider specifying the "fcst_thresh" for the CTC line type.
```
Comment (jprestop) [inline met/docs/Users_Guide/stat-analysis.rst]:
```suggestion
6. For the CTC line type, set the "column" entry to the CTS output column that contains the statistic of intereest (e.g. PODY for probability of detecting yes). Note, consider specifying the "fcst_thresh" for the CTC line type.
```

File: met/docs/Users_Guide/stat-analysis.rst [hunk 3]
```diff
@@ -52,10 +52,49 @@ When aggregating the matched pair line type (MPR) and computing an output contin
...
+
+For each term, all matching SL1L2 (or CTC) input lines are aggregated separtely for the forecast and reference models. The requested statistic ("column") is derived from the aggregated partial sums or counts. For each term, a skill score is defined as:
```
Comment (jprestop) [inline met/docs/Users_Guide/stat-analysis.rst]:
```suggestion
For each term, all matching SL1L2 (or CTC) input lines are aggregated separately for the forecast and reference models. The requested statistic ("column") is derived from the aggregated partial sums or counts. For each term, a skill score is defined as:
```

**Review by jprestop — APPROVED** [review]
I reviewed the extensive changes both in GitHub and on ReadTheDocs to ensure the formatting was correct.  I reviewed the log files at kiowa:/d1/projects/MET/MET_pull_requests/met-10.1.0/met-10.1.0_beta2/feature_1788/test_feature_1788_SSIDX/log and found no errors.  I reviewed the documentation for grammar/spelling, but don't have the knowledge to review the functionality.  I approve this request.

**Comment by JohnHalleyGotway** [comment]
@michelleharrold I found a problem. The computed GO Index value has changed when it should not have, as seen below:

In kiowa:d1/projects/MET/MET_pull_requests/met-10.1.0/met-10.1.0_beta2/feature_1788/pull_request_1892
```
MET-develop/met/bin/stat_analysis -lookin MET-develop/test_output/ref_config/point_stat \
-job go_index -model AFWAv3.4_Noahv3.3,AFWAv3.4_Noahv2.7.1 \
-vx_mask FULL -interp_mthd BILIN -v 4 -log go_index_develop.log
COL_NAME: GO_INDEX
GO_INDEX:  0.99873
```
```
MET-feature_1788_SSIDX/met/bin/stat_analysis -lookin MET-feature_1788_SSIDX/test_output/ref_config/point_stat \
-job go_index -model AFWAv3.4_Noahv3.3,AFWAv3.4_Noahv2.7.1 \
-vx_mask FULL -interp_mthd BILIN -v 4 -log go_index_feature_1788.log
COL_NAME:        FCST_MODEL           REF_MODEL N_INIT N_TERM N_VLD SS_INDEX
 GO_INDEX AFWAv3.4_Noahv3.3 AFWAv3.4_Noahv2.7.1      1     48    48  1.00254
```
I will investigate further... I found fixed the bug as noted in this [commit](https://github.com/dtcenter/MET/commit/50d3e38e9b664f4933ba3b84df19b66fb1016f7a) message. The feature branch now does produce the expected result:
```
MET-feature_1788_SSIDX/met/bin/stat_analysis -lookin MET-feature_1788_SSIDX/test_output/ref_config/point_stat \
-job go_index -model AFWAv3.4_Noahv3.3,AFWAv3.4_Noahv2.7.1 \
-vx_mask FULL -interp_mthd BILIN -v 4 -log go_index_feature_1788.log
COL_NAME:        FCST_MODEL           REF_MODEL N_INIT N_TERM N_VLD SS_INDEX
 GO_INDEX AFWAv3.4_Noahv3.3 AFWAv3.4_Noahv2.7.1      1     48    48  0.99873
```

**Comment by michelleharrold** [comment]
@JohnHalleyGotway - I tested this extensively with the executable on kiowa: /d1/projects/MET/MET_pull_requests/met-10.1.0/met-10.1.0_beta2/feature_1788/MET-feature_1788_SSIDX/met/bin/stat_analysis. I used the input data you provided here: /d1/projects/MET/MET_pull_requests/met-10.1.0/met-10.1.0_beta2/feature_1788/test_feature_1788_SSIDX

I ran several tests for CBS index, GO index, and customizing the indices via the StatAnalysisConfig file and exercised several of the command line options (-by, -set_hdr). All tests worked as expected. I will note that some of the data for the CBS index calculation is missing (specifically for valid time: 20181210_120000). The scores are still computed, but warnings are provided stating that certain terms on the index are missing. I am torn on whether there should be some sort of valid data criteria, or if the onus should be on the user to know their data. Thoughts?

For the documentation, I have a few minor comments:

Under [14.2.5 Skill Score Index](https://met.readthedocs.io/en/feature_1788_ssidx/Users_Guide/stat-analysis.html#skill-score-index), in the first sentence, I would take out 'pressure' and leave it to say 'meteorological fields at different levels' to keep it generic for AGL, etc.

Under [14.2.7 CBS Index](https://met.readthedocs.io/en/feature_1788_ssidx/Users_Guide/stat-analysis.html#cbs-index), should 'defined by a default Stat-Analysis configuration file' be 'defined by the default STATAnalysisConfig_CBS_Index configuration file'?

**Comment by JohnHalleyGotway** [comment]
@michelleharrold thanks for the testing! I just made the 2 doc updates you recommended.

As for dealing with missing data, I'm not sure. Options are:

(1) Leave it as-is with warnings for missing terms and reporting the number of requested terms and the number of valid terms in the N_TERM and N_VLD columns, respectively.

(2) Add an ss_index_vld_thresh = (0 to 1) config option to specify the required ratio of N_VLD / N_TERM in order to produce an output index value. ss_index_vld_thresh = 0.5 would require that you have data for at least 1/2 the terms. ss_index_vld_thresh = 1.0 would require that all data be present. We use that vld_thresh logic in lots of other places in the code.

(3) Add support for a new array of booleans to specify whether or not each term is required.
```
   ss_index_req = TRUE; // for all terms being required
   ss_index_req = FALSE; // for no specific term being required
   ss_index_req = [ TRUE, TRUE, FALSE, TRUE ]; // for only the 3rd of 4 terms being required
```
We do this in TC-Pairs when defining a consensus forecast, listing which models are or are not required.

Which of those 3 options makes the most sense here?

**Comment by michelleharrold** [comment]
I am torn on the missing data, but to consider how other user's might want to use this, I would err on adding some logic for missing data. I would choose option #3 -- but want to see if @hertneky has any input?

**Comment by hertneky** [comment]
Would option 3 just account for missing terms (I understand terms as HGT,SLP, WIND, etc). What about if forecast initializations are missing - would we just leave that as a warning?

**Comment by JohnHalleyGotway** [comment]
The code currently produces two types of warnings that we should think about.
(1) Not enough data for computing the skill score for a given term:
```
WARNING: SSIndexJobInfo::compute_ss_index() -> can't compute the skill score for term 39 (WIND/P250 at 960000 lead).
```
(2) The number of forecast and reference inputs do not match for a given term:
```
WARNING: SSIndexJobInfo::compute_ss_index() -> the number of aggregated forecast and reference lines differ (30 != 28) for term 21 (WIND/P250 at 240000 lead).
```

What I've proposed would only apply to type (1), where we CANNOT compute one of the terms. You could get warning type (2) if computing CBS Index over a month and 2 days were missing from the reference but not the forecast. For type (2), we ARE able to compute the skill score using 30 input forecast lines and 28 input reference lines.

What should the logic and config options be?

**Comment by hertneky** [comment]
@JohnHalleyGotway Thanks for the clarification. I understand now. The 3rd option is more complete in customizing missing data allowances. Setting each term to True/False seems a bit tedious and I am trying to think of how that would be useful. I could see a user not having certain input (i.e. forecasts not out to f120, or no variable at a certain level), but would it be best to just omit that from their config file rather than determining which terms those are to set to False?

**Comment by JohnHalleyGotway** [comment]
@hertneky and @michelleharrold, here's how I think this could be useful. Let's say that you have a month of output from point_stat and/or grid_stat and you want to compute skill score indices from it... separately for each model initialization. And let's say that you have some sporadic missing data during the month for whatever reason. Stat-Analysis computes the skill score indices using the data that is available which results in 28 scores where N_VLD = N_TERM and 2 scores where N_VLD < N_TERM. But all 30 scores are written to the same .stat output file. If you load that into METviewer, you'll have all 30 scores regardless of how N_VLD compares to N_TERM.

If we had 'ss_index_vld_thresh = 1.0' we'd only write output for the 28 days where N_VLD = N_TERM and the dataset wouldn't be "polluted" by the cases with missing data. To me, specifying the required ratio of valid data seems simple and useful... mostly as a way to avoid cases with ANY missing data (assuming that's desirable).

@hertneky you're right. Editing an array of 40 or 48 TRUE/FALSE values is tedious and error prone.

So if we do make a change, personally, I think option (2) is useful and sufficient. But I'll defer to what you guys decide... although I'm hoping you can decide this afternoon! So I can get it updated, tested, committed and included in the beta2 release tomorrow.

**Comment by michelleharrold** [comment]
I still like option 3 best because it allows more flexibility, but option 2 is certainly sufficient. I am fine with option 2.

**Comment by JohnHalleyGotway** [comment]
@michelleharrold you're paying the bills. I'll proceed with 3 and setting "ss_index_term_req = TRUE;" is just as easy as setting "ss_index_vld_thresh = 1.0;".  What should the default be?
```
ss_index_term_req = TRUE or FALSE?
```

**Comment by michelleharrold** [comment]
Haha, even though I am footing the bill, I still appreciate a democratic approach! Since you and Tracy have raised good points for number 2, I think it is best to go with the majority in this case. I am, therefore, officially switching to option 2. For that, I would prefer having the default be 'ss_index_vld_thresh = 1.0'

**Comment by JohnHalleyGotway** [comment]
OK, I added the ss_index_vld_thresh option. To confirm, this thresholds the ratio N_VLD to N_TERM... which addresses the first issue (where there is no data for one of the terms).

But it does NOT affect the case where the number of input forecast lines does not match the number of input reference lines. If we want to be more strict, we could require that n_fcst = n_ref and if they do not, consider that term to be bad data... and therefore subject to the ss_index_vld_thresh option.

But I will not make that change unless @michelleharrold and/or @hertneky instruct me to.

This is just a reminder of how messy data is! Especially when aggregating over many cases!

**Comment by michelleharrold** [comment]
@JohnHalleyGotway - thanks for adding in the ss_index_vld_thresh. I can envision a situation where having the functionality of n_fcst = n_ref would be handy, but I don't think it is necessary. The ss_index_vld_thresh option with no other functionality for handling n_fcst = n_ref checks all of our boxes, so let's stick with it. Thanks!

**Review by michelleharrold — APPROVED** [review]
I approve this PR. Thanks, John!

**Comment by JohnHalleyGotway** [comment]
@michelleharrold and @hertneky, if you approve of these changes, please complete the review process:
- Go to the "Files changed" tab for this PR.
- On the right side you'll see a little green button that either says "Review changes" or "Add your review".
- Click that, select the "Approve" radio button and any comment you'd like.
- Then click the "Submit Review" button.
- That puts a little green checkmark next to your name on this PR indicating your approval ;)

**Pull request merged**